src/main/scala/com/fulcrumgenomics/bam/AddReadGroupsByName.scala

```diff
@@ -0,0 +1,182 @@
+          if (readGroup.getPlatformModel != platformModel) unreachable(s"PM mismatch: s'${readGroup.getPlatformModel} != $platformModel")
+
+          // These should be set across all reads, so should never happen!
+          if (platform.exists(_ != readGroup.getPlatform)) unreachable(s"PL mismatch: s'${readGroup.getPlatform} != ${platform.get}")
+          if (sequencingCenter.exists(_ != readGroup.getSequencingCenter)) unreachable(s"CN mismatch: s'${readGroup.getSequencingCenter} != sequencingCenter.get}")
+          if (predictedInsertSize.exists(_ != readGroup.getPredictedMedianInsertSize)) unreachable(s"PI mismatch: s'${readGroup.getPredictedMedianInsertSize} != ${predictedInsertSize.get}")
+          if (description.exists(_ != readGroup.getDescription)) unreachable(s"DS mismatch: s'${readGroup.getDescription} != ${description.get}")
+          if (runDate.exists(_ != readGroup.getRunDate)) unreachable(s"DT mismatch: s'${readGroup.getRunDate} != ${runDate.get}")
+        }
+        progress.record(record)
+      }
+      in.safelyClose()
+    }
+
+    // Write them all out
+    {
+      val progress = new ProgressLogger(logger, verb = "written", unit = 5e6.toInt)
+      val in = SamReaderFactory.make().open(input.toFile)
+      val header = in.getFileHeader
+      val outHeader = header.clone()
+      outHeader.setSortOrder(sortOrder)
+      outHeader.setReadGroups(readGroups.values.toSeq)
+      comments.foreach(outHeader.addComment)
+      val out = new SAMFileWriterFactory().makeWriter(outHeader, sortOrder == header.getSortOrder, output.toFile, null)
+
+      in.iterator().foreach { record =>
+        val runInfo = RunInfo(name = record.getReadName)
+        val readGroupId = readGroups(runInfo).getReadGroupId
+        record.setAttribute(SAMTag.RG.name(), readGroupId)
+        out.addAlignment(record)
+        progress.record(record)
+      }
+
+      in.safelyClose()
+      out.close()
+    }
+  }
+
+  /** Generates a read group identifier for the read: <sample>.<platform-unit>. */
+  private def toReadGroupId(runInfo: RunInfo): String = s"$sample.${runInfo.platformUnit}"
+}
+
+/** Stores parsed run-level information about a read. */
+private[bam] case class RunInfo(instrument: String,
+                                runNumber: String,
+                                flowcellId: String,
+                                lane: Int)
+{
+  /** The platform unit: <flowcell.lane> */
+  val platformUnit: String = s"$flowcellId.$lane"
+
+  /** The platform mode: <instrument.run-number> */
+  def platformModel: String = s"$instrument.$runNumber"
+}
+
+private[bam] object RunInfo {
+  /** Parse a read name to get run-level information. */
+  def apply(name: String): RunInfo = {
+    val tokens = name.split(":").toSeq
+    if (tokens.length != 7) throw new IllegalArgumentException(s"Expected seven colon-delimited fields for read with name: $name")
```

Comment:
Back on line 46 the format includes a bunch more `:` separated fields after a space, so if you were to get the stated format this would fail.  Do you want to split on spaces first?